src/wasm/wasm-binary.cpp

```diff
@@ -1319,6 +1319,32 @@ Expression* WasmBinaryBuilder::popExpression() {
   return ret;
 }
 
+Expression* WasmBinaryBuilder::popNonVoidExpression() {
+  auto* ret = popExpression();
+  if (ret->type != none) return ret;
+  // we found a void, so this is stacky code that we must handle carefully
+  Builder builder(wasm);
+  // add elements until we find a non-void
+  std::vector<Expression*> expressions;
+  expressions.push_back(ret);
+  while (1) {
+    auto* curr = popExpression();
+    expressions.push_back(curr);
+    if (curr->type != none) break;
+  }
+  auto* block = builder.makeBlock();
+  while (!expressions.empty()) {
```

Comment:
Could you just do `std::copy_backward(expressions.begin(), expressions.end(), block->list.end())`